File: release-artifacts/v4-external-evidence-handoff/fixtures/external-engine-baselines/v4/unity/V4ExternalAssetImportWorkflowRunner.cs
// Galileo3D V4 Unity asset-import workflow evidence runner.
// Copy this file into a real Unity project under Assets/Galileo3D/V4ExternalBaselines.
// Run in batchmode with:
//   -executeMethod V4ExternalAssetImportWorkflowRunner.CaptureFromCommandLine --asset <path-to-gltf-or-glb> --evidence <repo>/tests/reports/v4-unity-asset-import-workflow.evidence.json
// The runner writes ok=false unless Unity actually imports the asset and exposes mesh/material/texture metrics.

#if UNITY_EDITOR
using System;
using System.IO;
using System.Linq;
using UnityEditor;
using UnityEngine;

public static class V4ExternalAssetImportWorkflowRunner
{
    public static void CaptureFromCommandLine()
    {
        string assetPath = Arg("--asset");
        string evidencePath = Arg("--evidence");
        if (string.IsNullOrEmpty(evidencePath))
        {
            throw new InvalidOperationException("Missing --evidence <path> for V4 Unity asset-import workflow evidence.");
        }
        Directory.CreateDirectory(Path.GetDirectoryName(evidencePath));
        var conversionRequired = new[] { "dae", "fbx", "usd", "usdz" };
        var nativeSupported = new[] { "glb", "gltf", "obj" };
        bool assetExists = !string.IsNullOrEmpty(assetPath) && File.Exists(assetPath);
        string unityAssetPath = assetExists ? CopyIntoAssets(assetPath) : "";
        if (assetExists)
        {
            AssetDatabase.ImportAsset(unityAssetPath, ImportAssetOptions.ForceUpdate | ImportAssetOptions.ImportRecursive);
            AssetDatabase.Refresh();
        }
        UnityEngine.Object[] imported = assetExists ? AssetDatabase.LoadAllAssetsAtPath(unityAssetPath) : Array.Empty<UnityEngine.Object>();
        int meshes = imported.Count((asset) => asset is Mesh);
        int materials = imported.Count((asset) => asset is Material);
        int textures = imported.Count((asset) => asset is Texture);
        int animationClips = imported.Count((asset) => asset is AnimationClip);
        bool ok = assetExists && meshes >= 1 && materials >= 1 && textures >= 1;
        string json = "{" +
            "\n  \"ok\": " + Bool(ok) + "," +
            "\n  \"engine\": \"unity\"," +
            "\n  \"workflowKind\": \"asset-import\"," +
            "\n  \"editorProjectOpened\": true," +
            "\n  \"assetImportWorkflowRan\": " + Bool(assetExists) + "," +
            "\n  \"assetPath\": " + Json(assetPath) + "," +
            "\n  \"importedFormats\": [\"glb\", \"gltf\"]," +
            "\n  \"nativeSupportedFormats\": [" + string.Join(", ", nativeSupported.Select(Json)) + "]," +
            "\n  \"conversionRequiredFormats\": [" + string.Join(", ", conversionRequired.Select(Json)) + "]," +
            "\n  \"metrics\": {" +
            "\n    \"editorProjectOpened\": true," +
            "\n    \"assetImportWorkflowRan\": " + Bool(assetExists) + "," +
            "\n    \"importedGltfAssets\": " + (assetExists ? 1 : 0) + "," +
            "\n    \"importedMeshes\": " + meshes + "," +
            "\n    \"importedMaterials\": " + materials + "," +
            "\n    \"importedTextures\": " + textures + "," +
            "\n    \"importedAnimationClips\": " + animationClips + "," +
            "\n    \"conversionRequiredFormats\": " + conversionRequired.Length + "," +
            "\n    \"nativeSupportedFormats\": " + nativeSupported.Length +
            "\n  }," +
            "\n  \"claimBoundary\": \"This sidecar is valid only when produced by a real Unity editor import run. It allows Galileo3D bounded native OBJ geometry import, but does not claim native FBX/USD/USDZ/DAE support.\"" +
            "\n}\n";
        File.WriteAllText(evidencePath, json);
        if (!ok)
        {
            throw new InvalidOperationException("Unity asset-import workflow did not expose enough imported mesh/material/texture metrics. Evidence written to " + evidencePath);
        }
    }

    private static string Arg(string name)
    {
        string[] args = Environment.GetCommandLineArgs();
        for (int index = 0; index < args.Length - 1; index++)
        {
            if (args[index] == name) return args[index + 1];
        }
        return "";
    }

    private static string CopyIntoAssets(string sourcePath)
    {
        string extension = Path.GetExtension(sourcePath);
        string target = "Assets/Galileo3D/V4ExternalBaselines/Imported/v4-import-workflow" + extension;
        Directory.CreateDirectory(Path.GetDirectoryName(target));
        File.Copy(sourcePath, target, true);
        return target;
    }

    private static string Bool(bool value) => value ? "true" : "false";
    private static string Json(string value) => "\"" + value.Replace("\\", "\\\\").Replace("\"", "\\\"") + "\"";
}
#endif

File: release-artifacts/v4-external-evidence-handoff/fixtures/external-engine-baselines/v4/unreal/v4_external_asset_import_workflow_runner.py
"""
Galileo3D V4 Unreal asset-import workflow evidence runner.

Run inside a real Unreal Editor Python session, for example:
  UnrealEditor-Cmd <project.uproject> -ExecutePythonScript="v4_external_asset_import_workflow_runner.py <asset-path> <repo>/tests/reports/v4-unreal-asset-import-workflow.evidence.json"

The script writes ok=false unless Unreal actually imports the asset and exposes
mesh/material/texture metrics. It allows Galileo3D bounded native OBJ geometry
import, but does not claim native FBX/USD/USDZ/DAE parity.
"""
import json
import os
import sys
import unreal

ASSET_PATH = sys.argv[1] if len(sys.argv) > 1 else ""
EVIDENCE_PATH = sys.argv[2] if len(sys.argv) > 2 else "tests/reports/v4-unreal-asset-import-workflow.evidence.json"
DESTINATION_PATH = "/Game/Galileo3D/V4AssetImportWorkflow"
CONVERSION_REQUIRED_FORMATS = ["dae", "fbx", "usd", "usdz"]
NATIVE_SUPPORTED_FORMATS = ["glb", "gltf", "obj"]

def import_asset(asset_path):
    if not asset_path or not os.path.exists(asset_path):
        return []
    task = unreal.AssetImportTask()
    task.filename = asset_path
    task.destination_path = DESTINATION_PATH
    task.automated = True
    task.save = True
    task.replace_existing = True
    unreal.AssetToolsHelpers.get_asset_tools().import_asset_tasks([task])
    return list(task.imported_object_paths or [])

def class_name(asset):
    try:
        return asset.get_class().get_name()
    except Exception:
        return ""

imported_paths = import_asset(ASSET_PATH)
loaded_assets = []
for object_path in imported_paths:
    try:
        asset = unreal.EditorAssetLibrary.load_asset(object_path)
        if asset is not None:
            loaded_assets.append(asset)
    except Exception as error:
        unreal.log_warning("Unable to load imported asset " + str(object_path) + ": " + str(error))

class_names = [class_name(asset) for asset in loaded_assets]
meshes = sum(1 for name in class_names if "Mesh" in name)
materials = sum(1 for name in class_names if "Material" in name)
textures = sum(1 for name in class_names if "Texture" in name)
animation_clips = sum(1 for name in class_names if "Anim" in name or "Animation" in name)
ok = bool(imported_paths) and meshes >= 1 and materials >= 1 and textures >= 1
evidence = {
    "ok": ok,
    "engine": "unreal",
    "workflowKind": "asset-import",
    "editorProjectOpened": True,
    "assetImportWorkflowRan": bool(ASSET_PATH and os.path.exists(ASSET_PATH)),
    "assetPath": ASSET_PATH,
    "importedObjectPaths": imported_paths,
    "importedFormats": ["glb", "gltf"],
    "nativeSupportedFormats": NATIVE_SUPPORTED_FORMATS,
    "conversionRequiredFormats": CONVERSION_REQUIRED_FORMATS,
    "metrics": {
        "editorProjectOpened": True,
        "assetImportWorkflowRan": bool(ASSET_PATH and os.path.exists(ASSET_PATH)),
        "importedGltfAssets": 1 if imported_paths else 0,
        "importedMeshes": meshes,
        "importedMaterials": materials,
        "importedTextures": textures,
        "importedAnimationClips": animation_clips,
        "conversionRequiredFormats": len(CONVERSION_REQUIRED_FORMATS),
        "nativeSupportedFormats": len(NATIVE_SUPPORTED_FORMATS),
    },
    "claimBoundary": "This sidecar is valid only when produced by a real Unreal editor import run. It allows Galileo3D bounded native OBJ geometry import, but does not claim native FBX/USD/USDZ/DAE support.",
}
directory = os.path.dirname(EVIDENCE_PATH)
if directory:
    os.makedirs(directory, exist_ok=True)
with open(EVIDENCE_PATH, "w", encoding="utf-8") as evidence_file:
    json.dump(evidence, evidence_file, indent=2)
    evidence_file.write("\n")
unreal.log("Galileo3D V4 Unreal asset-import workflow evidence written: " + EVIDENCE_PATH)
if not ok:
    raise RuntimeError("Unreal asset-import workflow did not expose enough imported mesh/material/texture metrics. Evidence written to " + EVIDENCE_PATH)
